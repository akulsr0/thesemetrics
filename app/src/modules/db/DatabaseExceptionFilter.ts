import { ArgumentsHost, Catch, ExceptionFilter } from '@nestjs/common';
import { FastifyReply } from 'fastify';
import { EntityNotFoundError } from 'typeorm/error/EntityNotFoundError';

@Catch(EntityNotFoundError)
export class DatabaseExceptionFilter implements ExceptionFilter {
  catch(exception: EntityNotFoundError, host: ArgumentsHost) {
    const ctx = host.switchToHttp();
    const response = ctx.getResponse<FastifyReply<Response>>();

    response.status(404).send({
      statusCode: 404,
      message: 'Not found',
    });
  }
}
