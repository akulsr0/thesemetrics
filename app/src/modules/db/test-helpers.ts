import { DynamicModule } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import {
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  PageView,
  Project,
  ProjectEvent,
  Team,
  User,
} from '@thesemetrics/database';
import FS from 'fs';
import Formatter from 'sql-formatter';

export interface SqlQuery {
  statement: string;
  parameters?: any[];
}

export interface SqlError extends SqlQuery {
  message: string;
}

export const SQL_ERRORS = '@test/sql-errors';
export const SQL_QUERIES = '@test/sql-queries';
export class TestDatabaseModule {
  static forRoot(): DynamicModule {
    const queries: SqlQuery[] = [];
    const errors: SqlError[] = [];

    return {
      module: TestDatabaseModule,
      imports: [
        TypeOrmModule.forRoot({
          type: 'sqljs',
          logging: true,
          synchronize: false,
          logger: {
            log: () => {},
            logMigration: () => {},
            logQuerySlow: () => {},
            logSchemaBuild: () => {},
            logQueryError: (message, statement, parameters) => errors.push({ message, statement, parameters }),
            logQuery: (statement, parameters) => queries.push({ statement, parameters }),
          },
          entities: [
            User,
            Team,
            Project,
            PageView,
            DailyAggregateDevice,
            DailyAggregatePageView,
            DailyAggregateReferrerPageView,
            ProjectEvent,
          ],
        }),
      ],
      providers: [
        {
          provide: SQL_QUERIES,
          useValue: queries,
        },
        {
          provide: SQL_ERRORS,
          useValue: errors,
        },
      ],
    };
  }
}

function stringify(data: any) {
  switch (typeof data) {
    case 'undefined':
      return '';
    case 'string':
      return `'${data.replace(/'/g, "\\'")}'`;
    case 'object':
      if (data instanceof Date) {
        return `'${data.toISOString()}'`;
      }
    default:
      return JSON.stringify(data);
  }
}

export const QuerySerializer = {
  serialize(queries: SqlQuery[]) {
    let content = '\n';

    Array.from(queries).forEach((query, index) => {
      content += `-- Query ${index + 1}\n`;
      content += Formatter.format(query.statement + ';', {
        language: 'sql',
        params: query.parameters ? query.parameters.map(stringify) : undefined,
      });
      content += '\n\n';
    });

    return content;
  },
  test(value: any) {
    return (
      Array.isArray(value) &&
      value.every((query) => {
        if (typeof query === 'object') {
          if ('statement' in query) {
            return (
              typeof query.statement === 'string' &&
              (typeof query.statement === 'undefined' || Array.isArray(query.parameters))
            );
          }
        }
        return false;
      })
    );
  },
};
