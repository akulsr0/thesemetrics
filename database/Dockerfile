FROM node:12-alpine

WORKDIR /app

ADD package.json /app/package.json
RUN npm install

ADD migrations/*.js /app/migrations/
