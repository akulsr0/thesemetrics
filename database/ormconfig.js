if (process.env.NODE_ENV !== 'production') {
  process.env.POSTGRES_URL = process.env.POSTGRES_URL || 'postgres://user:pass@localhost:5432/analytics';
}

module.exports = {
  type: 'postgres',
  url: process.env.POSTGRES_URL,
  ssl:
    process.env.NODE_ENV !== 'production'
      ? false
      : {
          ca: Buffer.from(process.env.POSTGRES_CERTIFICATE),
        },
  entities: [],
  migrations: ['migrations/*.js'],
  cli: {
    migrationsDir: 'migrations',
  },
};
