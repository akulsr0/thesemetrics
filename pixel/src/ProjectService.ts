import { Injectable } from '@nestjs/common'
import { InjectRepository } from '@nestjs/typeorm'
import {
  Project,
  ProjectEvent,
  ProjectType,
  Team
} from '@thesemetrics/database'
import { Repository } from 'typeorm'

export interface ProjectEventPayload
  extends Omit<ProjectEvent, 'id' | 'createdAt' | 'project'> {
  domain: string
}

@Injectable()
export class ProjectService {
  constructor(
    @InjectRepository(Team)
    private readonly teams: Repository<Team>,
    @InjectRepository(Project)
    private readonly projects: Repository<Project>,
    @InjectRepository(ProjectEvent)
    private readonly events: Repository<ProjectEvent>
  ) {}

  public async isDomainAllowed(domain: string): Promise<boolean> {
    if (!(await this.projects.count({ where: { domain } }))) {
      const team = await this.teams.findOne(1)

      await this.projects.save({
        team,

        type: ProjectType.WEBSITE,
        name: domain,
        domain: domain,
      })
    }

    return true
  }

  public async mayBePushEvent(event: ProjectEventPayload): Promise<void> {
    if (!(await this.isDomainAllowed(event.domain))) return // -- ignore event.

    const project = await this.projects.findOneOrFail({
      where: { domain: event.domain },
    })

    console.log(event)

    await this.events.save({
      ...event,
      project,
    })
  }
}
