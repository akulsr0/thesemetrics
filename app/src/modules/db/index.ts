import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import {
  User,
  Team,
  Project,
  PageView,
  DailyAggregateDevice,
  DailyAggregatePageView,
  DailyAggregateReferrerPageView,
  ProjectEvent,
  config,
} from '@thesemetrics/database';
import { APP_FILTER } from '@nestjs/core';
import { DatabaseExceptionFilter } from '@/modules/db/DatabaseExceptionFilter';

export * from '@thesemetrics/database';

@Module({
  imports: [
    TypeOrmModule.forRoot({
      ...config(),
      entities: [
        User,
        Team,
        Project,
        PageView,
        DailyAggregateDevice,
        DailyAggregatePageView,
        DailyAggregateReferrerPageView,
        ProjectEvent,
      ],
      synchronize: false,
    }),
  ],
  providers: [{ provide: APP_FILTER, useClass: DatabaseExceptionFilter }],
})
export default class DatabaseModule {}
