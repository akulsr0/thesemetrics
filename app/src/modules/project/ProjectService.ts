import { InjectRepository } from '@nestjs/typeorm';
import { Project } from '@/modules/db';
import { Repository } from 'typeorm';

export class ProjectService {
  constructor(@InjectRepository(Project) private readonly projects: Repository<Project>) {}

  async findProjectByDomain(domain: string) {
    return this.projects.findOneOrFail({
      where: { domain },
    });
  }
}
