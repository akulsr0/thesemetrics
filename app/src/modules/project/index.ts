import { Module } from '@nestjs/common';
import { ProjectService } from './ProjectService';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Project } from '@thesemetrics/database';
import DatabaseModule from '@/modules/db';

export { ProjectService } from './ProjectService';

@Module({
  imports: [DatabaseModule, TypeOrmModule.forFeature([Project])],
  providers: [ProjectService],
  exports: [ProjectService],
})
export default class ProjectModule {}
