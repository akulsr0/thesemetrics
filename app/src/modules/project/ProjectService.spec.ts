import { QuerySerializer, SqlQuery, TestDatabaseModule, SQL_QUERIES } from '@/modules/db/test-helpers';
import { ProjectService } from '@/modules/project/ProjectService';
import { Test } from '@nestjs/testing';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Project } from '@thesemetrics/database';

describe('ProjectService', () => {
  let service: ProjectService;
  let queries: SqlQuery[];

  expect.addSnapshotSerializer(QuerySerializer);

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        TestDatabaseModule.forRoot(),
        TypeOrmModule.forFeature([Project]),
      ],
      providers: [ProjectService],
    }).compile();

    service = moduleRef.get(ProjectService);
    queries = moduleRef.get(SQL_QUERIES)
  });

  describe('findProjectByDomain', () => {
    it('should query project by domain name', async () => {
      await expect(service.findProjectByDomain('example.com')).rejects.toThrow();
      expect(queries).toMatchSnapshot();
    });
  });
});
