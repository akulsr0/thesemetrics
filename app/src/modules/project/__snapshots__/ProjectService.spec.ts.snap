// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectService findProjectByDomain should query project by domain name 1`] = `

-- Query 1
SELECT
  "Project"."id" AS "Project_id",
  "Project"."name" AS "Project_name",
  "Project"."type" AS "Project_type",
  "Project"."domain" AS "Project_domain",
  "Project"."preferences" AS "Project_preferences",
  "Project"."createdAt" AS "Project_createdAt",
  "Project"."updatedAt" AS "Project_updatedAt",
  "Project"."teamId" AS "Project_teamId"
FROM
  "project" "Project"
WHERE
  "Project"."domain" = 'example.com'
LIMIT
  1;


`;
